top-level-walruses/piet/generator.py

```diff
@@ -0,0 +1,221 @@
+import base64
+import math
+import sys
+from enum import Enum
+from io import BytesIO
+from typing import Iterator
+
+from PIL import Image
+
+from .common import BLACK, PIET_COLORS, WHITE, Color, ColorChange, OrderedPair, SelfExpandingList
+from .interpreter import PietInterpreter
+from .runtime import DirectionOffset, DirectionPointer, PointerDirection
+
+
+class PietCommand(Enum):
+    _PREVIOUS = -1
+    _NONE = 0
+    NOOP = Color(255, 255, 255)
+    BLOCK = Color(0, 0, 0)
+    PUSH = ColorChange(1, 0)
+    POP = ColorChange(2, 0)
+    ADD = ColorChange(0, 1)
+    SUBTRACT = ColorChange(1, 1)
+    MULTIPLY = ColorChange(2, 1)
+    DIVIDE = ColorChange(0, 2)
+    MOD = ColorChange(1, 2)
+    NOT = ColorChange(2, 2)
+    GREATER = ColorChange(0, 3)
+    POINTER = ColorChange(1, 3)
+    SWITCH = ColorChange(2, 3)
+    DUPLICATE = ColorChange(0, 4)
+    ROLL = ColorChange(1, 4)
+    IN_NUMBER = ColorChange(2, 4)
+    IN_CHAR = ColorChange(0, 5)
+    OUT_NUMBER = ColorChange(1, 5)
+    OUT_CHAR = ColorChange(2, 5)
+
+
+class ImageGenerator:
+    def __init__(
+        self,
+        *,
+        input: str | bytes = b"",
+        step_limit: int = 1_000_000,
+        debug: bool = False,
+    ):
+        self.input = input
+        self.step_limit = step_limit
+        self.debug = debug
+        self.commands = SelfExpandingList(default=SelfExpandingList[PietCommand](default=PietCommand._NONE))
+        self.colors = SelfExpandingList(default=SelfExpandingList(default=BLACK))
+        self.interpreter = PietInterpreter(self.image, input=input, step_limit=step_limit, debug=debug)
+        self.interpreter.runtime.output = open(sys.stdout.fileno(), "wb", closefd=False)
+        self._current_hue = 0
+        self._current_lightness = 0
+        self._previous_color = WHITE
+
+    def set_command(self, command: PietCommand, position: OrderedPair):
+        y, x = position
+        self.commands[y][x] = command
+        if isinstance(command.value, ColorChange):
+            if self._previous_color not in (PietCommand.BLOCK.value, PietCommand.NOOP.value):
+                self._current_hue = (self._current_hue + command.value.hue) % 6
+                self._current_lightness = (self._current_lightness + command.value.lightness) % 3
+            color = PIET_COLORS[self._current_lightness][self._current_hue]
+        elif isinstance(command.value, Color):
+            color = command.value
+        else:
+            color = WHITE
+        self.colors[y][x] = color
+        self._previous_color = color
+
+    def set_offset_command(self, command: PietCommand, offset: DirectionOffset):
+        pointer = DirectionPointer()
+        pointer.direction = self.interpreter.runtime.pointer.direction
+        pointer.position = self.interpreter.runtime.pointer.position
```

Comment:
Similar to the `set_command` function, this function needs to know the current position of the internal interpreter, so it can pre-determine the next location of the pointer and place the command where it will be executed next. It needs a temporary pointer to simulate moving to the offset position before placing the command. The offset is useful for modifying commands not in the next position of the pointer, but offset in one of 8 directions (N, E, S, W, NE, NW, SE, SW). For example, placing black codels ahead of the next position in order to force a specific execution path.